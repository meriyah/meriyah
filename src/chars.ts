/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
