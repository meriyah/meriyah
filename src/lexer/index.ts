export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
