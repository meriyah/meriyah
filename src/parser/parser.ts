import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
