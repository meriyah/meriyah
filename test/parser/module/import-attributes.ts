import * as t from 'node:assert/strict';
import { outdent } from 'outdent';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail, pass } from '../../test-utils';

describe('Next - Import Attributes', () => {
  for (const arg of [
    "import 'bar' with { type: 'what' };",
    "import foo from 'bar' with { type: 'json' };",
    "import foo from 'bar' with { type: 'json', 'data-type': 'json' };",
    "import foo from 'bar' with { \"type\": 'json' };",
    "import foo from 'bar' with { \"type\": 'json', 'data-type': 'json' };",
    "import {default as viaStaticImport2} from './json-idempotency_FIXTURE.json' with { type: 'json' };",
    "import {default as viaStaticImport2} from './json-idempotency_FIXTURE.json' with { \"type\": 'json' };",
    "import * as ns from './json-via-namespace_FIXTURE.json' with { type: 'json' };",
    "import * as ns from './json-via-namespace_FIXTURE.json' with { \"type\": 'json' };",
    "import { random } from './random.ts' with { type: 'macro' };",
    "import { random } from './random.ts' with { \"type\": 'macro' };",
    "import { random as foo } from './random.ts' with { type: 'macro' };",
    "import { \"random\" as foo } from './random.ts' with { type: 'macro' };",
    "import('module', { type: 'json' });",
    "import('module', { lazy: true });",
    "import('module', { dynamic: false });",
    "import('module', { 'data-type': 'json' });",
    "import('module', { 'data-version': '1.0' });",
    "import('module', { type: 'text/javascript', version: '2.0' });",
    "import('module', { 'type': 'module', 'import-attr': true });",
    "'use strict'; import('module', { type: 'json' });",
    "async function load() { return import('module', { type: 'json' }); }",
    "(() => import('module', { type: 'json' }))()",
    "let dynamicImport = import('module', { type: 'json' });",
    "let dynamicImport = import('module', { \"type\": 'json' });",
    "const importWithAttributes = import('module', { type: 'json' });",
    "({ async load() { return import('module', { type: 'json' }); } })",
    "({ async load() { return import('module', { \"type\": 'json' }); } })",
    "function* gen() { yield import('module', { type: 'json' }); }",
    "function* gen() { yield import('module', { \"type\": 'json' }); }",
    "for await (let module of [import('module', { type: 'json' })]) {}",
    "for await (let module of [import('module', { \"type\": 'json' })]) {}",
    "import('module', { type: 'json' }).then(module => { /* ... */ });",
    "import('module', { 'data-type': 'json' }).then(module => { /* ... */ });",
    "const result = await import('module', { type: 'json' });",
    "const result = await import('module', { \"type\": 'json' });",
    "import x from './import-attribute-1_FIXTURE.js' with {};",
    "import './import-attribute-2_FIXTURE.js' with {};",
    "export * from './import-attribute-3_FIXTURE.js' with {};",
    "(async function () { return import('./2nd-param_FIXTURE.js', await undefined);}())",
    "var iter = function*() { beforeCount += 1, import('', yield), afterCount += 1;}();",
    "var promise; for (promise = import('./2nd-param_FIXTURE.js', 'test262' in {} || undefined); false; );",
    "export {default as viaStaticImport2} from './json-idempotency_FIXTURE.json' with { type: 'json' };",
    "export {default as viaStaticImport2} from './json-idempotency_FIXTURE.json' with { \"type\": 'json' };",
    "export * as foo from './foo.json' with { \"type\": 'json' };",
    "export * from './foo.json' with { \"type\": 'json' };",
    "export { random } from './random.ts' with { type: 'macro' };",
    "export { random } from './random.ts' with { \"type\": 'macro' };",
    "import 'bar' with { type: 'json' };",
    "import {} from 'bar' with { type: 'json' };",
    'import a, { foo } from "./foo.json" with { type: "json" };',
    'import { default as a, foo } from "./foo.json" with { type: "json" };',
    "export { foo } from './foo.json' with { type: 'json' };",
    "import {name} from './json-named-bindings_FIXTURE.json' with { type: 'json' };",
    "import {name} from './json-named-bindings_FIXTURE.json' with { \"type\": 'json' };",
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { sourceType: 'module' });
      });
    });
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { sourceType: 'module', webcompat: true });
      });
    });
  }

  fail('Expressions - Import Attributes (fail)', [
    'import("module", { type: "json" }, "extra")',
    'import("module", { type: "json" }, "extra")',
    { code: 'import("module", { type: "json", "extra": })', options: { sourceType: 'module' } },
    { code: 'import("module", ...extra)', options: { sourceType: 'module' } },
    { code: 'import("module", { type: "json", "extra": "value" ', options: { sourceType: 'module' } },
    { code: 'import("module", { type: "json", "extra": "value" }, "another")', options: { sourceType: 'module' } },
    { code: 'import("module", { type: "json" }, "extra")', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: "json", "data-type": "json"', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: ', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: "json", ', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: "json", "data-type": ', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: "json", "data-type": "json" ', options: { sourceType: 'module' } },
    {
      code: outdent`
        import x from './import-attribute-1_FIXTURE.js' with {
          type: 'json',
          'typ\u0065': ''
        };
      `,
      options: { sourceType: 'module' },
    },
    { code: 'import foo from "bar" with { 1: "foo" };', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: 1 };', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: [1] };', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: null };', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: undefined };', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: "json", foo: {} };', options: { sourceType: 'module' } },
    { code: "export foo, { foo2 } from './foo.json' with { \"type\": 'json' };", options: { sourceType: 'module' } },
  ]);

  pass('Import Attributes (pass)', [
    { code: "import('module', { type: 'json' });", options: { sourceType: 'module' } },
    { code: "import('module', { 'data-type': 'json' });", options: { sourceType: 'module' } },
    { code: "async function load() { return import('module', { type: 'json' }); }", options: { sourceType: 'module' } },
    { code: "for await (let module of [import('module', { type: 'json' })]) {}", options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: "json" };', options: { sourceType: 'module' } },
    { code: 'import foo from "bar" with { type: "json", "data-type": "json" };', options: { sourceType: 'module' } },
    {
      code: "var promise; for (promise = import('./2nd-param_FIXTURE.js', 'test262' in {} || undefined); false; );",
      options: { sourceType: 'module' },
    },
    { code: "export * from './foo' with { type: 'json' }", options: { sourceType: 'module' } },
    { code: "export * as foo from './foo' with { type: 'json' };", options: { sourceType: 'module' } },
    { code: "export {} from './foo' with { type: 'html' };", options: { sourceType: 'module' } },
    { code: "export { foo } from './foo' with { type: 'html' }", options: { sourceType: 'module' } },
    { code: "export { foo, } from './foo' with { type: 'html' };", options: { sourceType: 'module' } },
  ]);
});
