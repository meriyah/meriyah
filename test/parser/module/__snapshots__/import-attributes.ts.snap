// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > export foo, { foo2 } from './foo.json' with { "type": 'json' }; 1`] = `
"SyntaxError [1:7-1:10]: Unexpected token: 'identifier'
> 1 | export foo, { foo2 } from './foo.json' with { "type": 'json' };
    |        ^^^ Unexpected token: 'identifier'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { 1: "foo" }; 1`] = `
"SyntaxError [1:29-1:30]: Unexpected token: 'number'
> 1 | import foo from "bar" with { 1: "foo" };
    |                              ^ Unexpected token: 'number'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type:  1`] = `
"SyntaxError [1:34-1:35]: Unexpected token: 'end of source'
> 1 | import foo from "bar" with { type: 
    |                                   ^ Unexpected token: 'end of source'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: "json",  1`] = `
"SyntaxError [1:42-1:43]: Unexpected token: 'end of source'
> 1 | import foo from "bar" with { type: "json", 
    |                                           ^ Unexpected token: 'end of source'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: "json", "data-type":  1`] = `
"SyntaxError [1:55-1:56]: Unexpected token: 'end of source'
> 1 | import foo from "bar" with { type: "json", "data-type": 
    |                                                        ^ Unexpected token: 'end of source'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: "json", "data-type": "json"  1`] = `
"SyntaxError [1:62-1:63]: Expected ','
> 1 | import foo from "bar" with { type: "json", "data-type": "json" 
    |                                                               ^ Expected ','"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: "json", "data-type": "json" 1`] = `
"SyntaxError [1:56-1:62]: Expected ','
> 1 | import foo from "bar" with { type: "json", "data-type": "json"
    |                                                         ^^^^^^ Expected ','"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: "json", foo: {} }; 1`] = `
"SyntaxError [1:48-1:49]: Unexpected token: '{'
> 1 | import foo from "bar" with { type: "json", foo: {} };
    |                                                 ^ Unexpected token: '{'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: [1] }; 1`] = `
"SyntaxError [1:35-1:36]: Unexpected token: '['
> 1 | import foo from "bar" with { type: [1] };
    |                                    ^ Unexpected token: '['"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: 1 }; 1`] = `
"SyntaxError [1:35-1:36]: Unexpected token: 'number'
> 1 | import foo from "bar" with { type: 1 };
    |                                    ^ Unexpected token: 'number'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: null }; 1`] = `
"SyntaxError [1:35-1:39]: Unexpected token: 'null'
> 1 | import foo from "bar" with { type: null };
    |                                    ^^^^ Unexpected token: 'null'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import foo from "bar" with { type: undefined }; 1`] = `
"SyntaxError [1:35-1:44]: Unexpected token: 'identifier'
> 1 | import foo from "bar" with { type: undefined };
    |                                    ^^^^^^^^^ Unexpected token: 'identifier'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import x from './import-attribute-1_FIXTURE.js' with {
  type: 'json',
  'type': ''
}; 1`] = `
"SyntaxError [4:0-4:1]: Duplicate binding 'type'
  2 |   type: 'json',
  3 |   'type': ''
> 4 | };
    | ^ Duplicate binding 'type'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import("module", ...extra) 1`] = `
"SyntaxError [1:17-1:20]: Unexpected token: '...'
> 1 | import("module", ...extra)
    |                  ^^^ Unexpected token: '...'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import("module", { type: "json" }, "extra") 1`] = `
"SyntaxError [1:35-1:42]: Expected ')'
> 1 | import("module", { type: "json" }, "extra")
    |                                    ^^^^^^^ Expected ')'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import("module", { type: "json" }, "extra") 2`] = `
"SyntaxError [1:35-1:42]: Expected ')'
> 1 | import("module", { type: "json" }, "extra")
    |                                    ^^^^^^^ Expected ')'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import("module", { type: "json" }, "extra") 3`] = `
"SyntaxError [1:35-1:42]: Expected ')'
> 1 | import("module", { type: "json" }, "extra")
    |                                    ^^^^^^^ Expected ')'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import("module", { type: "json", "extra": "value"  1`] = `
"SyntaxError [1:49-1:50]: Expected '}'
> 1 | import("module", { type: "json", "extra": "value" 
    |                                                  ^ Expected '}'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import("module", { type: "json", "extra": "value" }, "another") 1`] = `
"SyntaxError [1:53-1:62]: Expected ')'
> 1 | import("module", { type: "json", "extra": "value" }, "another")
    |                                                      ^^^^^^^^^ Expected ')'"
`;

exports[`Next - Import Attributes > Expressions - Import Attributes (fail) > import("module", { type: "json", "extra": }) 1`] = `
"SyntaxError [1:42-1:43]: Unexpected token: '}'
> 1 | import("module", { type: "json", "extra": })
    |                                           ^ Unexpected token: '}'"
`;

exports[`Next - Import Attributes > Import Attributes (pass) > async function load() { return import('module', { type: 'json' }); } 1`] = `
{
  "body": [
    {
      "async": true,
      "body": {
        "body": [
          {
            "argument": {
              "options": {
                "properties": [
                  {
                    "computed": false,
                    "key": {
                      "name": "type",
                      "type": "Identifier",
                    },
                    "kind": "init",
                    "method": false,
                    "shorthand": false,
                    "type": "Property",
                    "value": {
                      "type": "Literal",
                      "value": "json",
                    },
                  },
                ],
                "type": "ObjectExpression",
              },
              "source": {
                "type": "Literal",
                "value": "module",
              },
              "type": "ImportExpression",
            },
            "type": "ReturnStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "generator": false,
      "id": {
        "name": "load",
        "type": "Identifier",
      },
      "params": [],
      "type": "FunctionDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > export * as foo from './foo' with { type: 'json' }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
      ],
      "exported": {
        "name": "foo",
        "type": "Identifier",
      },
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "type": "ExportAllDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > export * from './foo' with { type: 'json' } 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
      ],
      "exported": null,
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "type": "ExportAllDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > export { foo } from './foo' with { type: 'html' } 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "html",
          },
        },
      ],
      "declaration": null,
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "specifiers": [
        {
          "exported": {
            "name": "foo",
            "type": "Identifier",
          },
          "local": {
            "name": "foo",
            "type": "Identifier",
          },
          "type": "ExportSpecifier",
        },
      ],
      "type": "ExportNamedDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > export { foo, } from './foo' with { type: 'html' }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "html",
          },
        },
      ],
      "declaration": null,
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "specifiers": [
        {
          "exported": {
            "name": "foo",
            "type": "Identifier",
          },
          "local": {
            "name": "foo",
            "type": "Identifier",
          },
          "type": "ExportSpecifier",
        },
      ],
      "type": "ExportNamedDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > export {} from './foo' with { type: 'html' }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "html",
          },
        },
      ],
      "declaration": null,
      "source": {
        "type": "Literal",
        "value": "./foo",
      },
      "specifiers": [],
      "type": "ExportNamedDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > for await (let module of [import('module', { type: 'json' })]) {} 1`] = `
{
  "body": [
    {
      "await": true,
      "body": {
        "body": [],
        "type": "BlockStatement",
      },
      "left": {
        "declarations": [
          {
            "id": {
              "name": "module",
              "type": "Identifier",
            },
            "init": null,
            "type": "VariableDeclarator",
          },
        ],
        "kind": "let",
        "type": "VariableDeclaration",
      },
      "right": {
        "elements": [
          {
            "options": {
              "properties": [
                {
                  "computed": false,
                  "key": {
                    "name": "type",
                    "type": "Identifier",
                  },
                  "kind": "init",
                  "method": false,
                  "shorthand": false,
                  "type": "Property",
                  "value": {
                    "type": "Literal",
                    "value": "json",
                  },
                },
              ],
              "type": "ObjectExpression",
            },
            "source": {
              "type": "Literal",
              "value": "module",
            },
            "type": "ImportExpression",
          },
        ],
        "type": "ArrayExpression",
      },
      "type": "ForOfStatement",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > import foo from "bar" with { type: "json" }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
      ],
      "source": {
        "type": "Literal",
        "value": "bar",
      },
      "specifiers": [
        {
          "local": {
            "name": "foo",
            "type": "Identifier",
          },
          "type": "ImportDefaultSpecifier",
        },
      ],
      "type": "ImportDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > import foo from "bar" with { type: "json", "data-type": "json" }; 1`] = `
{
  "body": [
    {
      "attributes": [
        {
          "key": {
            "name": "type",
            "type": "Identifier",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
        {
          "key": {
            "type": "Literal",
            "value": "data-type",
          },
          "type": "ImportAttribute",
          "value": {
            "type": "Literal",
            "value": "json",
          },
        },
      ],
      "source": {
        "type": "Literal",
        "value": "bar",
      },
      "specifiers": [
        {
          "local": {
            "name": "foo",
            "type": "Identifier",
          },
          "type": "ImportDefaultSpecifier",
        },
      ],
      "type": "ImportDeclaration",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > import('module', { 'data-type': 'json' }); 1`] = `
{
  "body": [
    {
      "expression": {
        "options": {
          "properties": [
            {
              "computed": false,
              "key": {
                "type": "Literal",
                "value": "data-type",
              },
              "kind": "init",
              "method": false,
              "shorthand": false,
              "type": "Property",
              "value": {
                "type": "Literal",
                "value": "json",
              },
            },
          ],
          "type": "ObjectExpression",
        },
        "source": {
          "type": "Literal",
          "value": "module",
        },
        "type": "ImportExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > import('module', { type: 'json' }); 1`] = `
{
  "body": [
    {
      "expression": {
        "options": {
          "properties": [
            {
              "computed": false,
              "key": {
                "name": "type",
                "type": "Identifier",
              },
              "kind": "init",
              "method": false,
              "shorthand": false,
              "type": "Property",
              "value": {
                "type": "Literal",
                "value": "json",
              },
            },
          ],
          "type": "ObjectExpression",
        },
        "source": {
          "type": "Literal",
          "value": "module",
        },
        "type": "ImportExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;

exports[`Next - Import Attributes > Import Attributes (pass) > var promise; for (promise = import('./2nd-param_FIXTURE.js', 'test262' in {} || undefined); false; ); 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "name": "promise",
            "type": "Identifier",
          },
          "init": null,
          "type": "VariableDeclarator",
        },
      ],
      "kind": "var",
      "type": "VariableDeclaration",
    },
    {
      "body": {
        "type": "EmptyStatement",
      },
      "init": {
        "left": {
          "name": "promise",
          "type": "Identifier",
        },
        "operator": "=",
        "right": {
          "options": {
            "left": {
              "left": {
                "type": "Literal",
                "value": "test262",
              },
              "operator": "in",
              "right": {
                "properties": [],
                "type": "ObjectExpression",
              },
              "type": "BinaryExpression",
            },
            "operator": "||",
            "right": {
              "name": "undefined",
              "type": "Identifier",
            },
            "type": "LogicalExpression",
          },
          "source": {
            "type": "Literal",
            "value": "./2nd-param_FIXTURE.js",
          },
          "type": "ImportExpression",
        },
        "type": "AssignmentExpression",
      },
      "test": {
        "type": "Literal",
        "value": false,
      },
      "type": "ForStatement",
      "update": null,
    },
  ],
  "sourceType": "module",
  "type": "Program",
}
`;
